source/common/http/http3_status_tracker.cc

```diff
@@ -0,0 +1,46 @@
+#include "common/http/http3_status_tracker.h"
+
+namespace Envoy {
+namespace Http {
+
+namespace {
+
+// Initially, HTTP/3 is be marked broken for 5 minutes.
+const std::chrono::minutes DefaultExpirationTime{5};
+
+} // namespace
+
+Http3StatusTracker::Http3StatusTracker(Event::Dispatcher& dispatcher)
+    : expiration_timer_(dispatcher.createTimer([this]() -> void { onExpirationTimeout(); })) {}
+
+bool Http3StatusTracker::isHttp3Broken() const { return state_ == State::Broken; }
+
+bool Http3StatusTracker::isHttp3Confirmed() const { return state_ == State::Confirmed; }
+
+void Http3StatusTracker::markHttp3Broken() {
+  state_ = State::Broken;
+  if (!expiration_timer_->enabled()) {
+    expiration_timer_->enableTimer(std::chrono::duration_cast<std::chrono::milliseconds>(
+        DefaultExpirationTime * (1 << consecutive_broken_count_)));
+    ++consecutive_broken_count_;
+  }
+}
+
+void Http3StatusTracker::markHttp3Confirmed() {
+  consecutive_broken_count_ = 0;
+  if (expiration_timer_->enabled()) {
+    expiration_timer_->disableTimer();
+  }
+  state_ = State::Confirmed;
+}
+
+void Http3StatusTracker::onExpirationTimeout() {
+  if (state_ != State::Broken) {
+    return;
```

Comment:
Can this early return ever happen?  I think it is safe to `ASSERT(state_ == State::Broken);` in this method.

Reply:
Done!

Reply:
Seems like this is not done? :)

How would this ASSERT play with the point made above about any sequence of broken <-> confirmed state transitions being possible?

Reply:
Heh, it was done but then we reverted it after deciding that we did need to support multiple simultaneous connection attempts.